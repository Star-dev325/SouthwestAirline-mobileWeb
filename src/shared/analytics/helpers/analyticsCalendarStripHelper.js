// @flow
import dayjs from 'dayjs';

import type { BoundType } from 'src/shared/flow-typed/flightBound.types';
import { INBOUND, OUTBOUND } from 'src/shared/constants/flightBoundTypes';

export const calculateDateDiffs = (
  newDate: string,
  previousDate: string,
  direction: BoundType,
  currentDiffs: { [BoundType]: string } = {}
): { [BoundType]: string } => {
  const diff = dayjs(newDate).diff(previousDate, 'days');
  const sign = Math.sign(diff) === 1 ? '+' : '';
  const diffs = { ...currentDiffs };

  diffs[direction] = `${sign}${diff}`;

  return diffs;
};

export const calculateOverrideDateDiffs = (
  newDate: string,
  previousDate: string,
  direction: BoundType,
  isOverrideEndDate: boolean,
  previousDiffs: { [BoundType]: string } = {}
): { [BoundType]: string } => {
  let diffs = {};

  if (direction === OUTBOUND) {
    if (isOverrideEndDate) {
      diffs = calculateDateDiffs(newDate, previousDate, INBOUND, previousDiffs);
    } else {
      diffs = {};
    }
  }

  return diffs;
};
