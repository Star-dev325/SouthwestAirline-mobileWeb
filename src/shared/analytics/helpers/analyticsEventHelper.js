import _ from 'lodash';

const pageIdentifierTransformer = [
  { matcher: /^\/$|^\/\?/, id: 'home' },
  { matcher: /\/flight-status\/(?!.*html).*/, id: 'flight-status-results' },

  { matcher: /\/flight-status\?.*_modal=originAirport/, id: 'flight-status-from' },
  { matcher: /\/flight-status\?.*_modal=destinationAirport/, id: 'flight-status-to' },

  { matcher: /\/air\/booking\/shopping\?.*_modal=from/, id: 'air-booking-shopping-from' },
  { matcher: /\/air\/booking\/shopping\?.*_modal=to/, id: 'air-booking-shopping-to' },
  { matcher: /\/air\/booking\/shopping\?.*_modal=airBookingCalendar/, id: 'air-booking-shopping-dates' },

  { matcher: /\/air\/change\/shopping\?.*_modal=from/, id: 'air-change-shopping-from' },
  { matcher: /\/air\/change\/shopping\?.*_modal=to/, id: 'air-change-shopping-to' },
  { matcher: /\/air\/change\/shopping\?.*_modal=airBookingCalendar/, id: 'air-change-shopping-dates' },

  { matcher: /\/car\/booking\?.*_modal=pickUpModal/, id: 'car-booking-pick-up' },
  { matcher: /\/car\/booking\?.*_modal=dropOffModal/, id: 'car-booking-drop-off' },
  { matcher: /\/car\/booking\?.*_modal=carBookingCalendar/, id: 'car-booking-dates' },

  { matcher: /\/check-in\/reservation\/\w{6}$/, id: 'checkin-view-reservation' }
];

const _matchPath = (path) =>
  _.get(
    _.find(pageIdentifierTransformer, (t) => path.match(t.matcher)),
    'id'
  );

export const transformPath = (path, search = null) => {
  if (!path) {
    return '';
  }
  let toMatchPath = path;

  if (search) {
    toMatchPath = `${path}${search}`;
  }

  return _matchPath(toMatchPath) || path.replace(/\//g, '-').slice(1);
};

export const raiseEvent = (eventName) => {
  window._trackAnalytics && window._trackAnalytics.event();

  const customEvent = new window.CustomEvent(eventName);

  document.body.dispatchEvent(customEvent);
};

export const raiseSatelliteEvent = (eventName, detail) => {
  if (window._satellite) {
    detail ? window._satellite.track(eventName, detail) : window._satellite.track(eventName);
  }
};

export const customerMessageAnalytics = (messages) => {
  if (messages) {
    const messageKeys = _.chain(messages).map('key').compact().join().value();

    _.set(window, 'data_a.message.customer', messageKeys);
    _.set(window, 'data_a.message.customerdisplay', '1');
  }
};

export const getPageIdentifier = (location) => {
  const { pathname = '', search = '' } = location || {};

  return transformPath(pathname, search);
};
