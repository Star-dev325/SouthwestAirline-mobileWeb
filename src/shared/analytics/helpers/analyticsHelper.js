import _ from 'lodash';

export const generateUpdatedFlowStoreForAnalytics = (flowSelectors, state, actionType) =>
  _.chain(flowSelectors)
    .pickBy((selectorObj) => _.includes(selectorObj.actions, actionType))
    .mapValues((selectorObj) => selectorObj.selector(state))
    .value();

export const generateFlowActionListForAnalytics = (flowSelectors) =>
  _.chain(flowSelectors).mapValues('actions').values().flatten().uniq().value();

export const generateUpdatedStoresForAnalytics = (generators, state, actionType) =>
  _.reduce(
    generators,
    (result, generator, key) => {
      _.each(generator(state, actionType), (subValue, subKey) => {
        result[`${key}.${subKey}`] = subValue;
      });

      return result;
    },
    {}
  );
