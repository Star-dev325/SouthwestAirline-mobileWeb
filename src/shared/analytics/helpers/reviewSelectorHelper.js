import _ from 'lodash';
import { transformToUnitPrice } from 'src/airBooking/transformers/transformToEarlyBirdPriceDetails';

export const buildReviewAnalyticsData = (
  earlyBirdEligibility,
  purposeOfTravel,
  earlyBirdInPathRadioButtonChecked,
  flightPricingPageAnalyticsData,
  earlyBirdSelected
) => ({
  earlyBirdBasePriceCents: _.get(earlyBirdEligibility, 'unitPrice.amount', null),
  earlyBirdEstimatedTotalCents: _.get(earlyBirdEligibility, 'totalPrice.amount', null),
  earlyBirdOptionSelected: earlyBirdInPathRadioButtonChecked || earlyBirdSelected,
  earlyBirdOptionShown: _.some(_.get(earlyBirdEligibility, 'bounds', []), 'isEligible'),
  purposeOfTravel,
  earlyBirdBasePriceOutbound: _.get(transformToUnitPrice(_.get(earlyBirdEligibility, 'bounds[0]')), 'amount', null),
  earlyBirdBasePriceInbound: _.get(transformToUnitPrice(_.get(earlyBirdEligibility, 'bounds[1]')), 'amount', null),
  ...flightPricingPageAnalyticsData
});
