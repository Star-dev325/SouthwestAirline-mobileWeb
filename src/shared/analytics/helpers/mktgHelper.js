// @flow
import _ from 'lodash';
import mktgDataConstants from 'src/shared/constants/mktgDataConstants';
import { createSelector } from 'src/shared/helpers/createSelectorWithErrorReporter';
import { getGlobalMktgData } from 'src/shared/analytics/selectors/getGlobalMktgData';
import { toNumberStringWithoutCommas } from 'src/shared/helpers/currencyValueHelper';
import type { FlightProductCard, ProductDefinition } from 'src/shared/flow-typed/shared.types';
import type { MKTGFareType } from 'src/shared/flow-typed/mktg.types';

const {
  NOT_AVAILABLE,
  NOT_SELECTED,
  DIRECT_FLIGHT,
  POINTS_BOOKING,
  CURRENCY_BOOKING,
  CONNECTING_FLIGHT,
  NONSTOP_FLIGHT
} = mktgDataConstants;

export const parseFareValue = (fare: MKTGFareType, fareValue: (fare: MKTGFareType) => string) =>
  (typeof fare === 'string' && fare) || (fare.reasonIfUnavailable && NOT_AVAILABLE) || fareValue(fare);

export const getFarePrice = (fare: MKTGFareType) => {
  const pricePath = fare.discountedPrice ? 'discountedPrice.amount' : 'price.amount';
  const farePrice = _.get(fare, pricePath, NOT_AVAILABLE);

  return (
    (farePrice === NOT_AVAILABLE && farePrice) ||
    (parseInt(farePrice, 10) === 0 && 'zero') ||
    toNumberStringWithoutCommas(farePrice)
  );
};

export const getFarePriceDifference = ({ priceDifference: { amount = NOT_AVAILABLE, sign = '' } = {} }: MKTGFareType) =>
  (amount === NOT_AVAILABLE && amount) ||
  (parseInt(amount) === 0 && 'zero') ||
  (sign === '-' && `${sign}${toNumberStringWithoutCommas(amount)}`) ||
  toNumberStringWithoutCommas(amount);

export const getSortedFareProducts = (products: Array<ProductDefinition>): Array<ProductDefinition> =>
  [...products].sort((product1, product2) => (product2.rowOrder > product1.rowOrder ? 1 : -1));

export const getSortedFares = (fares: Array<MKTGFareType>, products: Array<ProductDefinition>): Array<MKTGFareType> =>
  products.map(
    ({ productId }) =>
      fares.find((fare) => typeof fare !== 'string' && fare._meta.fareProductId === productId) || NOT_AVAILABLE
  );

export const getFareData = (
  propPrefix: string,
  fares: Array<MKTGFareType>,
  products: Array<ProductDefinition>,
  bookingType: string,
  fareValue: (MKTGFareType) => string = getFarePrice
) =>
  products.reduce((fareData, product, index) => {
    const fare = fares.find(
      (fareItem) => typeof fareItem !== 'string' && fareItem._meta.fareProductId === product.productId
    );
    const {
      [`${propPrefix}_viewfaretypes`]: fareTypes,
      [`${propPrefix}_viewfareproductids`]: fareIds,
      [`${propPrefix}_viewfarecurrency`]: fareCurrencyList,
      [`${propPrefix}_viewfarepoints`]: farePointsList
    } = fareData;
    const fareValueResult =
      (fare && parseFareValue(fare, fareValue)) ||
      (fares.every((fareItem) => fareItem === NOT_SELECTED) && NOT_SELECTED) ||
      NOT_AVAILABLE;
    const fareCurrency = bookingType === CURRENCY_BOOKING ? fareValueResult : NOT_SELECTED;
    const farePoints = bookingType === POINTS_BOOKING ? fareValueResult : NOT_SELECTED;
    const { id: fareType = NOT_AVAILABLE, productId: fareProductId = NOT_AVAILABLE } = product;
    const dataIndex = index + 1;

    return {
      ...fareData,
      [`${propPrefix}_viewfaretypes`]: fareTypes ? `${fareTypes}|${fareType}` : fareType,
      [`${propPrefix}_viewfareproductids`]: fareIds ? `${fareIds}|${fareProductId}` : fareProductId,
      [`${propPrefix}_viewfarecurrency`]: fareCurrencyList ? `${fareCurrencyList}|${fareCurrency}` : fareCurrency,
      [`${propPrefix}_viewfarepoints`]: farePointsList ? `${farePointsList}|${farePoints}` : farePoints,
      [`${propPrefix}_viewfareproductid${dataIndex}`]: fareType,
      [`${propPrefix}_viewfaretype${dataIndex}`]: fareType,
      [`${propPrefix}_viewfarecurrency${dataIndex}`]: fareCurrency,
      [`${propPrefix}_viewfarepoints${dataIndex}`]: farePoints
    };
  }, {});

export const getFareDifferencesData = (propPrefix1: string, propPrefix2: string, fareDifferences: Array<string>) =>
  fareDifferences.reduce(
    (result, difference, index) => ({
      ...result,
      [`${propPrefix1}_viewfarediff${propPrefix2}${index + 1}`]: difference
    }),
    {}
  );

export const getStopData = (propPrefix: string, flight: FlightProductCard) => {
  const numberOfStops = _.get(flight, '_meta.numberOfStops', NOT_SELECTED);
  const hasStop = parseInt(numberOfStops) > 0;
  const hasPlaneChange = !!flight.stopCity;

  const stopType = hasPlaneChange ? CONNECTING_FLIGHT : hasStop ? DIRECT_FLIGHT : NONSTOP_FLIGHT;

  return {
    [`${propPrefix}_stops`]: numberOfStops,
    [`${propPrefix}_stoptype`]: numberOfStops === NOT_SELECTED ? NOT_SELECTED : stopType
  };
};

export const getMktgDataForPage = (path: string, defaultValue: *) => (state: *) => _.get(state, path, defaultValue || {});

export const createMktgDataSelector = (path: string, defaultValue: *) => createSelector(
  [getGlobalMktgData, getMktgDataForPage(path, defaultValue)],
  (globalMktgData, pageMktgData) => ({ ...globalMktgData, ...pageMktgData })
);
