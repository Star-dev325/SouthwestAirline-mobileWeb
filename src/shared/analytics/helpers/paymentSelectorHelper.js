import _ from 'lodash';

import { getCardType } from 'src/shared/helpers/creditCardHelper';

import {
  NEW_CREDIT_CARD_ID,
  RAPID_REWARDS_VISA_ID,
  PAY_PAL_CARD_ID,
  APPLE_PAY_CARD_ID,
  UPLIFT_CARD_ID
} from 'src/shared/constants/creditCardConstants';
import { INSTANT_CREDIT_RAPID_REWARDS_VISA, PAYPAL, APPLE_PAY, UATP } from 'src/shared/constants/creditCardTypes';
import { addCurrency } from 'src/shared/api/helpers/currencyHelper';
import { isCurrencyAmountZero } from 'src/shared/helpers/travelFundsHelper';

export const buildPaymentAnalyticsData = (
  paymentInfo,
  savedCreditCards,
  flightConfirmationPage,
  travelFundsBalanceRemaining
) => {
  let typeOfCreditCard;
  const { city, stateProvinceRegion, zipOrPostalCode, saveCreditCard, cardNumber } = paymentInfo;
  const fundsApplied = _.get(flightConfirmationPage, 'fundsApplied');
  let selectedCardId = _.get(paymentInfo, 'selectedCardId');
  const selectedSavedCreditCard =
    _.find(_.chain(savedCreditCards).values().flatten().value(), { savedCreditCardId: selectedCardId }) || {};
  const creditCardType = getCardType(cardNumber);
  const primaryCardType = _.get(savedCreditCards, 'primaryCard.type');
  const rtfs = [];
  const luvVouchers = [];
  const giftCards = [];
  const isApplePay = _.isEqual(selectedCardId, APPLE_PAY_CARD_ID);
  const isUplift = _.isEqual(selectedCardId, UPLIFT_CARD_ID);

  if (!_.isEmpty(fundsApplied)) {
    let fundTypes = [];

    if (!_.isEmpty(paymentInfo) && !isCurrencyAmountZero(travelFundsBalanceRemaining)) {
      if (_.isEmpty(creditCardType) && selectedSavedCreditCard.type) {
        fundTypes = [selectedSavedCreditCard.type];
      } else if (isApplePay) {
        fundTypes = [APPLE_PAY.shortName];
      } else if (isUplift) {
        fundTypes = [UATP.shortName];
      } else if (_.isEmpty(creditCardType)) {
        fundTypes = [INSTANT_CREDIT_RAPID_REWARDS_VISA.key];
      } else {
        fundTypes = [creditCardType];
      }
    } else if (_.isEmpty(paymentInfo) && !isCurrencyAmountZero(travelFundsBalanceRemaining)) {
      fundTypes = [primaryCardType];
    }

    _.forEach(fundsApplied, (fund) => {
      const fundType = _.get(fund, 'travelFundType');
      const appliedAmount = _.get(fund, 'appliedAmount');

      !_.includes(fundTypes, fundType) && fundTypes.push(fundType);
      fundType === 'TRAVEL_FUNDS' && rtfs.push(appliedAmount);
      fundType === 'LUV_VOUCHER' && luvVouchers.push(appliedAmount);
      fundType === 'GIFT_CARD' && giftCards.push(appliedAmount);

      typeOfCreditCard = fundTypes.join(',');
    });
  } else {
    if (_.isEmpty(paymentInfo)) {
      typeOfCreditCard = isApplePay ? APPLE_PAY.shortName : primaryCardType;
      typeOfCreditCard = isUplift ? UATP.shortName : typeOfCreditCard;
      selectedCardId = _.get(savedCreditCards, 'primaryCard.savedCreditCardId');
    } else {
      typeOfCreditCard = _.isEmpty(creditCardType)
        ? selectedSavedCreditCard.type || INSTANT_CREDIT_RAPID_REWARDS_VISA.key
        : creditCardType;

      typeOfCreditCard = isApplePay ? APPLE_PAY.shortName : typeOfCreditCard;
      typeOfCreditCard = isUplift ? UATP.shortName : typeOfCreditCard;
    }
  }

  const type = selectedCardId === PAY_PAL_CARD_ID ? PAYPAL.key : typeOfCreditCard;

  return _.omitBy(
    {
      city,
      state: stateProvinceRegion,
      zipcode: zipOrPostalCode,
      saveNewCardSelected: _.toBoolean(saveCreditCard) || false,
      storedCard:
        selectedCardId &&
        !isCurrencyAmountZero(travelFundsBalanceRemaining) &&
        !_.includes(
          [NEW_CREDIT_CARD_ID, RAPID_REWARDS_VISA_ID, PAY_PAL_CARD_ID, APPLE_PAY_CARD_ID, UPLIFT_CARD_ID],
          selectedCardId
        ),
      type,
      travelfund:
        rtfs.length > 1
          ? _.get(addCurrency(...rtfs), 'amount')
          : !_.isEmpty(rtfs[0])
            ? _.get(rtfs[0], 'amount')
            : undefined,
      luvvoucher:
        luvVouchers.length > 1
          ? _.get(addCurrency(...luvVouchers), 'amount')
          : !_.isEmpty(luvVouchers[0])
            ? _.get(luvVouchers[0], 'amount')
            : undefined,
      giftcard:
        giftCards.length > 1
          ? _.get(addCurrency(...giftCards), 'amount')
          : !_.isEmpty(giftCards[0])
            ? _.get(giftCards[0], 'amount')
            : undefined
    },
    _.isUndefined
  );
};
