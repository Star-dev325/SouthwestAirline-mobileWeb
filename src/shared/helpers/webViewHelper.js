// @flow
import { handleRouteChange as handleRouteChangeFn } from '@swa-ui/hybrid';
import _ from 'lodash';
import { matchPath } from 'react-router';
import { history } from 'src/appHistory';
import * as LocalStorageCache from 'src/shared/cache/localStorageCache';
import appConfig from 'src/shared/config/appConfig';
import { DEBOUNCE_DELAY } from 'src/shared/constants/timeoutConstants';
import {
  appId,
  LOGIN_TYPES,
  LOGIN_STATES,
  NATIVE_FUNCTIONS,
  NORMALIZED_ROUTES,
  QUERY_PARAMS,
  ROUTES
} from 'src/shared/constants/webViewConstants';
import { isObject } from 'src/shared/helpers/jsUtils';
import { transformSearchToQuery } from 'src/shared/helpers/pathUtils';
import StorageKeys from 'src/shared/helpers/storageKeys';
import {
  getCurrentRouteState,
  getPrevRouteState,
  isBrowserBackOrForward
} from 'src/shared/routeUtils/routeStateHelper';
import store2 from 'store2';

import type { RecentSearchRequestType } from 'src/flightStatus/flow-typed/flightStatus.types';
import type { CurrencyType, NativeAppLoginOptions } from 'src/shared/flow-typed/shared.types';

const {
  WEB_VIEW_CHANNEL,
  WEB_VIEW_API_KEY,
  WEB_VIEW_CORPORATE_CHANNEL,
  OAUTH_LOGIN_STATUS,
  WEB_VIEW_EXPERIENCE_ID,
  WEB_VIEW_APP_VERSION
} = StorageKeys;
const { PENDING, NATIVE_LOG_IN, NATIVE_LOG_OUT } = LOGIN_STATES;
const {
  CHANNEL,
  CORPORATE_CHANNEL,
  WEB_VIEW,
  API_KEY,
  DEVICE_TYPE,
  SHARE_FLIGHT_STATUS,
  ADOBE_ID,
  EXPERIENCE_ID,
  APP_VERSION,
  UPGRADE_TYPE
} = QUERY_PARAMS;
const {
  BLANK,
  FLIGHT_STATUS,
  CAR_BOOKING,
  AIR_BOOKING,
  AIR_BOOKING_APPLY_RAPID_REWARDS,
  AIR_BOOKING_LEGACY,
  LOOKUP_TRAVEL_FUNDS,
  CONTACT_TRACING,
  OFFERS,
  UPGRADED_BOARDING,
  UPGRADED_BOARDING_PURCHASE,
  UPGRADED_BOARDING_CONFIRMATION,
  AIR_UPGRADE,
  AIR_UPGRADE_SELECT_BOUNDS,
  MY_ACCOUNT_PROMO_CODES
} = ROUTES;
const {
  AIR_BOOKING_INDEX,
  AIR_BOOKING_FLIGHT_SHOPPING
} = NORMALIZED_ROUTES;
const {
  EXIT,
  PAGE_RENDERED,
  DISPLAY_LOGIN,
  LOGOUT,
  UPDATE_FLIGHT_STATUS_SEARCHES,
  ENABLE_NAVIGATION_CONTROLS,
  DISPLAY_APP_REVIEW,
  SHARE_FLIGHT_STATUS_DETAILS,
  DISPLAY_APPLE_PAY,
  CHASE_PROMO_CLICKED,
  SAVE_CHASE_OFFERS
} = NATIVE_FUNCTIONS;

type Props = {
  isWebView: boolean,
  webViewLoginStatus: string
};

export const isWebViewLogin = (props: Props, prevProps: Props) => {
  const { isWebView, webViewLoginStatus } = props;
  const { webViewLoginStatus: prevWebViewLoginStatus } = prevProps;

  const didDisplayLogin = prevWebViewLoginStatus === PENDING;
  const didLogin = webViewLoginStatus === NATIVE_LOG_IN;

  return isWebView && didDisplayLogin && didLogin;
};

export const isWebViewLogout = (props: Props, prevProps: Props) => {
  const { isWebView, webViewLoginStatus } = props;
  const { webViewLoginStatus: prevWebViewLoginStatus } = prevProps;

  const didDisplayLogin = prevWebViewLoginStatus === PENDING;
  const didLogout = webViewLoginStatus === NATIVE_LOG_OUT;

  return isWebView && didDisplayLogin && didLogout;
};

const getQueryParam = (param: string) => {
  const queryParamsStr = _.get(history, 'location.search', '');
  const queryParams = transformSearchToQuery(queryParamsStr) || {};

  return _.get(queryParams, param);
};

export const hasWebViewInterface = () =>
  _.hasAny(window, [androidInterface(PAGE_RENDERED), iosInterface(PAGE_RENDERED)]);

export const getBooleanQueryParam = (paramKey: string) => getQueryParam(paramKey) === 'true';
export const hasWebViewParam = () => getBooleanQueryParam(WEB_VIEW);
export const getChannelParam = () => getQueryParam(CHANNEL);
export const getCorporateChannelParam = () => getQueryParam(CORPORATE_CHANNEL);
export const getApiKeyParam = () => getQueryParam(API_KEY);
export const getDeviceTypeParam = () => getQueryParam(DEVICE_TYPE);
export const getShareFlightStatusParam = () => getBooleanQueryParam(SHARE_FLIGHT_STATUS);
export const getUpgradeTypeParam = () => getQueryParam(UPGRADE_TYPE);
export const getAdobeIdParam = () => getQueryParam(ADOBE_ID);
export const getExperienceId = () => getQueryParam(EXPERIENCE_ID);
export const getAppVersion = () => getQueryParam(APP_VERSION);

export const isOnHomePage = () => _.get(history, 'location.pathname') === '/';

const androidInterface = (FUNCTION_NAME: string) => `AndroidInterface.${FUNCTION_NAME}`;
const iosInterface = (FUNCTION_NAME: string) => `webkit.messageHandlers.${FUNCTION_NAME}.postMessage`;
const simulatorInterface = (FUNCTION_NAME: string) => `SimulatorInterface.${FUNCTION_NAME}`;

const sendMessageToNativeApps = (FUNCTION_NAME: string, value: string) => {
  _.invoke(window, androidInterface(FUNCTION_NAME), value);
  _.invoke(window, iosInterface(FUNCTION_NAME), value);
  _.invoke(window, simulatorInterface(FUNCTION_NAME), value);
};

export const exitWebView = (route: string) => {
  sendMessageToNativeApps(EXIT, route);
  _.isEmpty(route) && handleRouteChangeFn(history, appId, BLANK, '');
};

export const chasePromoClicked = (target: string, linkType: string, isChaseCombo: boolean, referrer: string) => {
  const message = encodeMessage({
    target,
    linkType,
    isChaseCombo,
    referrer
  });

  sendMessageToNativeApps(CHASE_PROMO_CLICKED, message);
};

export const showNativeAppLogin = (options: NativeAppLoginOptions) => {
  const defaultOptions = { loginType: LOGIN_TYPES.NORMAL, continueAsGuest: false };
  const message = encodeMessage({ ...defaultOptions, ...options });

  sendMessageToNativeApps(DISPLAY_LOGIN, message);
};

export const showNativeApplePay = (moneyTotal: CurrencyType) =>
  sendMessageToNativeApps(DISPLAY_APPLE_PAY, encodeMessage(moneyTotal));

export const updateFlightStatusSearches = (flightStatusSearches: Array<RecentSearchRequestType>) =>
  sendMessageToNativeApps(UPDATE_FLIGHT_STATUS_SEARCHES, encodeMessage(flightStatusSearches));

export const shareFlightStatusDetails = (shareDetails: *) =>
  sendMessageToNativeApps(SHARE_FLIGHT_STATUS_DETAILS, encodeMessage(shareDetails));

export const sendChaseOffers = (chaseOffers: *) =>
  sendMessageToNativeApps(SAVE_CHASE_OFFERS, encodeMessage(chaseOffers));

export const enableNavigationControls = (isEnabled: boolean) =>
  sendMessageToNativeApps(ENABLE_NAVIGATION_CONTROLS, _.toString(isEnabled));

export const displayAppReview = () => sendMessageToNativeApps(DISPLAY_APP_REVIEW, '');

export const handleNativeLogout = () => sendMessageToNativeApps(LOGOUT, '');

const remember = (key: string, value: string) => {
  value && store2.set(key, value);
};

export const rememberChannel = (nativeChannel: string) => remember(WEB_VIEW_CHANNEL, nativeChannel);
export const rememberCorporateChannel = (nativeCorporateChannel: string) =>
  remember(WEB_VIEW_CORPORATE_CHANNEL, nativeCorporateChannel);
export const rememberApiKey = (nativeApiKey: string) => remember(WEB_VIEW_API_KEY, nativeApiKey);
export const rememberExperienceId = (nativeExperienceId: string) =>
  remember(WEB_VIEW_EXPERIENCE_ID, nativeExperienceId);
export const rememberAppVersion = (nativeAppVersion: string) => remember(WEB_VIEW_APP_VERSION, nativeAppVersion);

export const rememberChaseInfo = (encodedValue: string) => {
  const value = decodeMessage(encodedValue);
  const { chaseSessionId } = value;

  if (chaseSessionId) {
    LocalStorageCache.saveSwaOffersIdentity();
    LocalStorageCache.saveChaseSessionId(chaseSessionId);
    LocalStorageCache.deleteChasePrequalOffers();
  }
};

const getValidRoute = (route: string = '') => {
  const baseRoute = route.split('?')[0];

  switch (baseRoute) {
    case FLIGHT_STATUS:
    case CAR_BOOKING:
    case CONTACT_TRACING:
    case AIR_BOOKING:
    case AIR_BOOKING_APPLY_RAPID_REWARDS:
    case AIR_BOOKING_LEGACY:
    case AIR_BOOKING_INDEX:
    case AIR_BOOKING_FLIGHT_SHOPPING:
    case LOOKUP_TRAVEL_FUNDS:
    case OFFERS:
    case UPGRADED_BOARDING:
    case UPGRADED_BOARDING_PURCHASE:
    case UPGRADED_BOARDING_CONFIRMATION:
    case AIR_UPGRADE:
    case AIR_UPGRADE_SELECT_BOUNDS:
    case MY_ACCOUNT_PROMO_CODES:
      return route;
    default:
      return BLANK;
  }
};

export const isBlankPage = (routeKey: string) => _.startsWith(routeKey, BLANK);
export const isLandingPage = (routeKey: string) => getValidRoute(routeKey) !== BLANK;

export const decodeMessage = (encodedMessage: string) => {
  try {
    const decodedMessage = Buffer.from(encodedMessage, 'base64').toString() || '{}';

    return JSON.parse(decodedMessage) || {};
  } catch (e) {
    return {};
  }
};

export const encodeMessage = (decodedMessage: *): string => {
  try {
    const stringifiedMessage = JSON.stringify(decodedMessage);

    return Buffer.from(stringifiedMessage).toString('base64');
  } catch (e) {
    return '';
  }
};

export const isIndexPage = (routePath: string, currentPathName: string) => (
  isObject(routePath) ?
    Object.values(routePath).some((path) => matchPath(currentPathName, { path: String(path), exact: true })?.isExact ?? false)
    : matchPath(currentPathName, { path: routePath, exact: true })?.isExact ?? false
);

export const isOnWebViewLandingPage = (persistentHistory: Array<*>, landingPagePath: string) => {
  const currentState = getCurrentRouteState(persistentHistory);
  const previousState = getPrevRouteState(persistentHistory);

  const currentPathName = currentState?.pathname;
  const previousPathName = previousState?.pathname;

  const isOnLandingPage = isIndexPage(landingPagePath, currentPathName);
  const isFromBlankPage = matchPath(previousPathName, { path: '/blank', exact: true });

  return !!isOnLandingPage && !!isFromBlankPage && !isBrowserBackOrForward(currentState);
};

export const shouldShowContinueAsGuest = (loginType: string) =>
  ![LOGIN_TYPES.POINTS, LOGIN_TYPES.TRANSFER_TRAVEL_FUNDS].includes(loginType);

const hasIosToken = () => {
  const { scope } = store2.get(OAUTH_LOGIN_STATUS) || {};

  return _.includes(scope, 'ios');
};

const getWebViewApiValue = (localStorageKey: string, appConfigValue: string) => {
  const localStorageValue = store2.get(localStorageKey);

  return hasIosToken() && !localStorageValue ? appConfigValue : localStorageValue;
};

export const debouncedFn = _.debounce((fn: *) => fn && fn(), DEBOUNCE_DELAY, { leading: true, trailing: false });
export const getWebViewApiKey = () => getWebViewApiValue(WEB_VIEW_API_KEY, appConfig.IOS_API_KEY);
export const getWebViewChannel = () => getWebViewApiValue(WEB_VIEW_CHANNEL, appConfig.IOS_API_CHANNEL);
export const getWebViewCorporateChannel = () =>
  getWebViewApiValue(WEB_VIEW_CORPORATE_CHANNEL, appConfig.IOS_API_CORPORATE_CHANNEL);
