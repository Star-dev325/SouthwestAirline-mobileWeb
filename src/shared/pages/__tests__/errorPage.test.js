jest.mock('src/shared/helpers/webViewHelper', () => ({
  exitWebView: jest.fn()
}));

jest.mock('@swa-ui/hybrid', () => ({
  isHybridEnabled: jest.fn()
}));

import React from 'react';
import { Provider } from 'react-redux';
import { fireEvent, render } from '@testing-library/react';
import { exitWebView } from 'src/shared/helpers/webViewHelper';
import { isHybridEnabled } from '@swa-ui/hybrid';
import ErrorPage from 'src/shared/pages/errorPage';
import createMockStore from 'test/unit/helpers/createMockStore';

describe('error page', () => {
  let mockResetErrorBoundary;
  const { location } = window;

  beforeEach(() => {
    jest.clearAllMocks();
    jest.resetAllMocks();
    mockResetErrorBoundary = jest.fn();
    delete window.location;
    window.location = { replace: jest.fn() };
  });

  afterEach(() => {
    window.location = location;
  });

  it('should match snapshot', () => {
    const { container } = createComponent();

    expect(container).toMatchSnapshot();
  });

  it('should return home when button is clicked', () => {
    isHybridEnabled.mockReturnValue(false);
    const { getByText } = createComponent();

    fireEvent.click(getByText('SHARED__ERROR_PAGE__BUTTON'));

    expect(window.location.replace).toHaveBeenCalledWith('/');
  });

  it('should exit webview when button is clicked in a webview', () => {
    isHybridEnabled.mockReturnValue(true);
    const { getByText } = createComponent();

    fireEvent.click(getByText('SHARED__ERROR_PAGE__WEBVIEW_BUTTON'));

    expect(exitWebView).toHaveBeenCalled();
  });

  it('should not return home when button is clicked in a webview', () => {
    isHybridEnabled.mockReturnValue(true);
    const { getByText } = createComponent();

    fireEvent.click(getByText('SHARED__ERROR_PAGE__WEBVIEW_BUTTON'));

    expect(window.location.replace).not.toHaveBeenCalled();
  });

  it('should reset error state when unmounted', () => {
    const { unmount } = createComponent();

    unmount();

    expect(mockResetErrorBoundary).toHaveBeenCalled();
  });

  const createComponent = () => {
    const store = createMockStore()({});
    const props = {
      error: 'mock error',
      errorInfo: { componentStack: 'mock error info' },
      resetErrorBoundary: mockResetErrorBoundary
    };

    return render(
      <Provider store={store}>
        <ErrorPage {...props} />
      </Provider>
    );
  };
});
