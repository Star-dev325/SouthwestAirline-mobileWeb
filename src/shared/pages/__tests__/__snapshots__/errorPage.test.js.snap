// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`error page should match snapshot 1`] = `
<div>
  <div
    class="error-page"
  >
    <div
      class="info-banner "
    >
      <div
        class="info-banner--alert"
      >
        <i
          class="icon icon_exclamation-circle error"
        />
      </div>
      <div
        class="info-banner-container"
      >
        <b
          class="info-banner-container--header"
        >
          SHARED__ERROR_PAGE__BANNER_TITLE
        </b>
        <p
          class="info-banner-container--body"
        >
          SHARED__ERROR_PAGE__BANNER_DESCRIPTION
        </p>
        <div
          class="info-banner-container--link"
        />
      </div>
    </div>
    <div
      class="error-page--heading"
    >
      SHARED__ERROR_PAGE__TITLE
    </div>
    <div
      class="error-page--body"
    >
      <div
        data-qa="observed-viewport"
      >
        <div
          class="error-page--placement"
        >
          <div
            class="shimmer"
          >
            <div
              class="shimmer--children"
            />
          </div>
        </div>
      </div>
      <div
        data-qa="observed-viewport"
      >
        <div
          class="error-page--placement"
        >
          <div
            class="shimmer"
          >
            <div
              class="shimmer--children"
            />
          </div>
        </div>
      </div>
      <div
        data-qa="observed-viewport"
      >
        <div
          class="error-page--placement"
        >
          <div
            class="shimmer"
          >
            <div
              class="shimmer--children"
            />
          </div>
        </div>
      </div>
      <button
        class="button button--fluid larger button--yellow"
        type="button"
      >
        SHARED__ERROR_PAGE__BUTTON
      </button>
    </div>
  </div>
</div>
`;
